File: 78-subsets/78-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans, cur = [], []
        
        def backtrack(i):
            if i < len(nums):
                cur.append(nums[i])
                backtrack(i+1); cur.pop(); backtrack(i+1)
            else:
                ans.append(cur[:])
            return ans
        
        return backtrack(0)

File: 78-subsets/78-subsets.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> cur;
    void dfs(int idx, vector<int>& nums) {
        if (idx >= nums.size()) {
            res.push_back(cur);
            return;
        }
        cur.push_back(nums[idx]);
        dfs(idx + 1, nums);
        cur.pop_back();
        dfs(idx + 1, nums);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        dfs(0, nums);
        return res;
    }
};